File: alextune/tune.go
package main

import (
	"context"
	"os/signal"
	"syscall"
	"time"
	"flag"
	"fmt"
	"io"
	"os"
	"os/exec"
	"strings"

	log "github.com/sirupsen/logrus"
)

var (
	speed = flag.Int("speed", 38400, "")
	dev   = flag.String("dev", "/dev/ttyUSB0", "")

	displayMap = map[byte]byte{
		'>': '-',
		'<': 'l',
		'@': ' ',
		'K': 'H',
		'M': 'N',
		'Q': 'O',
		'V': 'U',
		'W': 'I',
		'X': '?', // c-bar
		'Z': 'c',
		'[': '?', // r-bar
		'\\': '?', // lambda
		']': '?', // RX/TX eq level 4
		'^': '?', // RX/TX eq level 4
	}
)

func command(f io.ReadWriter, cmd string, reply bool) (string, error) {
	_, err := f.Write([]byte(cmd))
	if err != nil {
		return "", err
	}
	if !reply {
		return "", nil
	}
	var ss []string
	for {
		b := make([]byte, 1, 1)
		n, err := f.Read(b)
		if err != nil {
			return "", err
		}
		s := string(b[:n])
		ss = append(ss, s)
		if s[len(s)-1] == ';' {
			return strings.Join(ss, ""), nil
		}
	}
}

func main() {
	flag.Parse()
	ctx := context.Background()
	if err := exec.CommandContext(ctx, "stty", "-F", *dev, "raw", fmt.Sprint(*speed)).Run(); err != nil {
		log.Fatalf("Setting terminal speed: %v", err)
	}
	f, err := os.OpenFile(*dev, os.O_RDWR|os.O_SYNC, 0)
	if err != nil {
		log.Fatalf("Opening serial port: %v", err)
	}
	defer f.Close()

	// Confirm we're using the right protocol.
	if id, err := command(f, "ID;", true);err != nil {
		log.Fatalf("Command failed: %v", err)
	} else if id != "ID017;" {
		log.Fatalf("Not protocol 17: %q", id)
	}

	// Confirm we're using a KX2.
	if id, err := command(f, "OM;", true);err != nil {
		log.Fatalf("Command failed: %v", err)
	} else if !strings.HasSuffix(id, "01;") {
		log.Fatalf("Not a KX2. OM: %q", id)
	}

	// Get old K2/K3 mode.
	oldK2, err := command(f, "K2;", true)
	if err != nil {
		log.Fatalf("Failed to get K2 mode: %v", err)
	}
	log.Printf("Old K2 mode: %q", oldK2)
	oldK3, err := command(f, "K3;", true)
	if err != nil {
		log.Fatalf("Failed to get K3 mode: %v", err)
	}
	log.Printf("Old K3 mode: %q", oldK3)

	signalCh := make(chan os.Signal)
	signal.Notify(signalCh, os.Interrupt, syscall.SIGTERM)

	// Set K2 extended mode.
	if r, err := command(f, "K22;K2;", true);err != nil {
		log.Fatalf("Command failed: %v", err)
	} else if false &&  r != "K20;" {
		log.Fatalf("Bad K20 reply: %q", r)
	}
	defer func() {
		if _, err := command(f, oldK2, false); err != nil {
			log.Fatalf("Failed to set old K2 mode %q: %v", oldK2, err)
		}
	}()

	// Set normal K3 mode.
	if r, err := command(f, "K31;K3;", true);err != nil {
		log.Fatalf("Command failed: %v", err)
	} else if false &&  r != "K30;" {
		log.Fatalf("Bad K30 reply: %q", r)
	}
	defer func() {
		if _, err := command(f, oldK3, false); err != nil {
			log.Fatalf("Failed to set old K2 mode %q: %v", oldK2, err)
		}
	}()

	// Get old power.
	oldPC, err := command(f, "PC;", true)
	if err != nil {
		log.Fatalf("Command failed: %v", err)
	}
	log.Printf("Old power: %q", oldPC)

	defer func() {
		// Set back old power.
		if _, err := command(f, oldPC, false); err != nil {
			log.Fatalf("Command failed: %v", err)
		}

		// Confirm power.
		if pc, err := command(f, "PC;", true); err != nil {
			log.Fatalf("Command failed: %v", err)
		} else if pc != oldPC {
			log.Fatalf("Power not actually set back to %q: %q", oldPC, pc)
		}
	}()

	// Get ATU status, and TX power.
	var oldATU bool
	if r, err := command(f, "DS;", true); err != nil {
		log.Fatalf("Command failed: %v", err)
	} else {
		if g,w:=len(r),13; g != w {
			log.Fatalf("DS reply wrong length. Got %d (%q), want %d.", g,r,w)
		}
		oldATU = (r[11] & 16) != 0
		log.Printf("ATU status: %v", oldATU)
	}

	// Turn off ATU and set power to 1W.
	if r, err := command(f, "MN023;MP001;MN255;PC0100;PC;", true); err != nil {
		log.Fatalf("Command failed: %v", err)
	} else {
		log.Printf("ATU off&power command return: %q", r)
	}

	// Confirm power.
	if pc, err := command(f, "PC;", true); err != nil {
		log.Fatalf("Command failed: %v", err)
	} else if pc != "PC0100;" {
		log.Fatalf("Power not actually set to 1W: %q", pc)
	}

	// Press TUNE.
	if _, err := command(f, "SWH16;", false); err != nil {
		log.Fatalf("Command failed: %v", err)
	}

	// Report SWR.
	log.Printf("Tune down SWR and press enter...\n")
	die := make(chan struct{})
	printLoopDone := make(chan struct{})
	go func() {
		defer close(printLoopDone)
		t := time.NewTicker(100*time.Millisecond)
		defer t.Stop()
		for {
			select {
			case <-die:
				fmt.Printf("\n")
				return
			case <-signalCh:
				fmt.Printf("\n")
				log.Infof("Got signal")
				return
			case <-t.C:
				o, err := command(f, "DS;", true)
				if err != nil {
					log.Fatalf("DS; failed: %v", err)
				}
				// Example: DS@@@@1³>I
				v := strings.TrimLeft(strings.TrimPrefix(o, "DS"), "@")
				out := ""
				bs := []byte(v)
				if len(bs) <3 {
					log.Errorf("Invalid DS reply: %q", o) 
				}
				for _, b := range bs[:len(bs)-3] {
					if b & 128  > 0{
						out += "."
						b &= 127
					}
					if b2, found := displayMap[b]; found {
						b = b2
					}
					out += fmt.Sprintf("%c", b)
				}
				fmt.Printf("\r                                              \rDS: %q (at %s)", out, time.Now().Format("2006-01-02 15:04:05"))
			}
		}
	}()

	// Wait for enter.
	go func() {
		defer close(die)
		b := make([]byte, 10, 10)
		if _, err := os.Stdin.Read(b); err != nil {
			log.Fatalf("Did you press ^D or something? %v", err)
		}
	}()
	<-printLoopDone
	
	// Turn off TUNE.
	if _, err := command(f, "RX;", false); err != nil {
		log.Fatalf("Command failed: %v", err)
	}

	if oldATU {
		log.Printf("Re-enabling ATU")
		time.Sleep(time.Second)
		// Turn on ATU.
		if _, err := command(f, "MN023;MP002;SWT09;", false); err != nil {
			log.Fatalf("Command failed: %v", err)
		}
	}

	if oldATU {
		// Tune ATU.
		if _, err := command(f, "SWT20;", false); err != nil {
			log.Fatalf("Command failed: %v", err)
		}
	}
	log.Infof("Done")
}

File: alextune/tune.cc
#include<errno.h>
#include<fcntl.h>
#include<iostream>
#include<map>
#include<poll.h>
#include<stdexcept>
#include<string.h>
#include<string>
#include<termios.h>
#include<unistd.h>
#include<vector>

void
setup(int fd, int speed)
{
	struct termios tio = {0};
	if (tcgetattr(fd, &tio) != 0) {
		throw std::runtime_error("tcgetattr");
	}

	if (cfsetospeed(&tio, speed)) {
		throw std::runtime_error("cfsetospeed");
	}
	if (cfsetispeed(&tio, speed)) {
		throw std::runtime_error("cfsetispeed");
	}

	// 8 Data bits.
	tio.c_cflag = (tio.c_cflag & ~CSIZE) | CS8;

	// No hardware control or parity or other fancy stuff.
	tio.c_iflag &= ~(IXON | IXOFF | IXANY | IGNBRK);
	tio.c_cflag |= CLOCAL;
	tio.c_cflag &= ~(PARENB | PARODD);
	tio.c_cflag &= ~CRTSCTS;
	tio.c_cflag |= CREAD;

	// 1 Stop bit.
	tio.c_cflag &= ~CSTOPB;
	tio.c_lflag = 0;
	tio.c_oflag = 0;
	tio.c_cc[VMIN]  = 0;
	tio.c_cc[VTIME] = 5;

	if (tcsetattr(fd, TCSANOW, &tio) != 0) {
		throw std::runtime_error("tcsetattr");
	}
}

std::string
command(int fd, const std::string& cmd, bool reply)
{
	//std::clog << "Running command: " << cmd << std::endl;
	{
		const char* p = cmd.data();
		int len = cmd.size();
		for (; len > 0;) {
			const ssize_t n = write(fd, p, len);
			if (n < 0) {
				throw std::runtime_error("write failure");
			}
			len -= n;
			p += n;
		}
		const auto n = write(fd, ";", 1);
		if (n != 1) {
			throw std::runtime_error("write failure");
		}
	}

	if (!reply) {
		return "";
	}
	std::string ret;
	for (;;) {
		char buf[1];
		int n = read(fd, buf, sizeof(buf));
		if (n < 0) {
			throw std::runtime_error("read failure");
		}
		if (n == 0) {
			throw std::runtime_error("EOF");
		}
		ret += std::string(&buf[0], &buf[n]);
		if (buf[n-1] == ';') {
			break;
		}
	}
	//std::clog << "... => " << ret << std::endl;
	return ret.substr(0,ret.size()-1);
}

bool
has_suffix(const std::string& s, const std::string& suffix)
{
	if (suffix.size() > s.size()) {
		return false;
	}
	return std::equal(suffix.rbegin(), suffix.rend(), s.rbegin());
}

std::string
trim_prefix(const std::string& s, const std::string& pre)
{
	if (s.substr(0, pre.size()) == pre) {
		return s.substr(pre.size(), s.size());
	}
	return s;
}

std::string
trim_left(const std::string& s)
{
	if (!s.empty() && s[0] == ' ') {
		return trim_left(s.substr(1, s.size()));
	}
	return s;
}

int
run()
{
	std::string portname = "/dev/ttyUSB0";
	int fd = open(portname.c_str(), O_RDWR | O_NOCTTY | O_SYNC);
	if (fd < 0) {
		throw std::runtime_error("open failed");
	}

	setup(fd, B38400);

	// Confirm radio.
	{
		// Confirm protocol.
		const auto id = command(fd, "ID", true);
		if (id != "ID017") {
			throw std::runtime_error("Not protocol 17: " + id);
		}

		// Confirm KX2.
		const auto om = command(fd, "OM", true);
		if (!has_suffix(om, "01")) {
			throw std::runtime_error("Not KX2: " + om);
		}
	}

	// Get power.
	const auto oldPC = command(fd, "PC", true);

	// Turn off ATU and set power to 1W.
	command(fd, "MN023;MP001;SWT09;PC001", false);

	// Confirm power.
	{
		const auto pc = command(fd, "PC", true);
		if (pc != "PC001") {
			throw std::runtime_error("Power not actually set to 1W: %q");
		}
	}

	// Press TUNE.
	command(fd, "SWH16", false);

	std::map<char, char> display_map {
		{	'>',  '-'},
		{		'<',  'l'},
		{		'@',  ' '},
		{		'K',  'H'},
		{		'M',  'N'},
		{		'Q',  'O'},
		{		'V',  'U'},
		{		'W',  'I'},
		{		'X',  '?'}, // c-bar
		{		'Z',  'c'},
		{		'[',  '?'}, // r-bar
		{		'\\', '?'}, // lambda
		{		']',  '?'}, // RX/TX eq level 4
		{		'^',  '?'}, // RX/TX eq level 4
			};

	printf("Tune down SWR and press enter...\n");
	for (;;) {
		struct pollfd fds;
		fds.fd = STDIN_FILENO;
		fds.events = POLLIN;
		if (poll(&fds, 1, 500)) {
			break;
		}
		const auto swr = command(fd, "DS", true);
		std::string out;
		for (int c = 0; c < swr.size(); c++) {
			auto b = swr[c];
			if (b & 128) {
				out.push_back('.');
				b &= 127;
			}
			const auto d = display_map.find(b);
			if (d != display_map.end()) {
				b = d->second;
			}
			out.push_back(b);
		}
		out = trim_prefix(out, "DS");
		out = trim_left(out);
		printf("\r                     \rDS: %s", out.c_str());
		fflush(stdout);
	}
	printf("\n");

	sleep(1);

	// Turn off TUNE.
	command(fd, "RX", false);

	sleep(1);

	// Turn on ATU.
	command(fd, "MN023;MP002;SWT09", false);

	// Set back old power.
	command(fd, oldPC, false);

	// Confirm power.
	{

		const auto pc = command(fd, "PC", true);
		if (pc != oldPC) {
			throw std::runtime_error("Power not actually set back to %q: %q");
		}
	}

	sleep(1);

	// Tune ATU.
	command(fd, "SWT20", false);
}

int
main()
{
	try {
		run();
	} catch(const std::exception&e) {
		std::cerr << e.what() << std::endl;
		return 1;
	}
	return 0;
}
